javaharness/java/arcs/api/PecPortManager.java

```diff
@@ -8,49 +8,66 @@
 @Singleton
 public class PecPortManager {
 
-  private final Map<String, PecMessageReceiver> ports = new HashMap<>();
-  private final PecInnerPortFactory portFactory;
+  private final ShellApi shellApi;
+  private final ParticleExecutionContext pec;
+  private final PortableJsonParser jsonParser;
+
+  private final Map<String, PecInnerPort> pecPortMap = new HashMap<>();
+  private final Map<String, PecInnerPortProxy> pecPortProxyMap = new HashMap<>();
 
   @Inject
-  PecPortManager(PecInnerPortFactory portFactory) {
-    this.portFactory = portFactory;
+  PecPortManager(
+    ShellApi shellApi,
+    ParticleExecutionContext pec,
+    PortableJsonParser jsonParser) {
+    this.shellApi = shellApi;
+    this.pec = pec;
+    this.jsonParser = jsonParser;
   }
 
   public void deliverPecMessage(String pecId, String sessionId, PortableJson message) {
-    PecMessageReceiver port = ports.get(pecId);
-    if (port == null) {
-      // Create a new inner PEC port for this pecId.
-      port = createInnerPecPort(pecId, sessionId);
+    PecInnerPort port = pecPortMap.get(pecId);
+    if (port != null) {
+      port.onReceivePecMessage(message);
+      return;
+    }
+
+    PecInnerPortProxy portProxy = pecPortProxyMap.get(pecId);
+    if (portProxy != null) {
+      portProxy.onReceivePecMessage(message);
+      return;
     }
 
+    // Create a new inner PEC port for this pecId.
+    port = createInnerPecPort(pecId, sessionId);
     port.onReceivePecMessage(message);
   }
 
-  public void addRemotePecPort(String pecId, RemotePecPort remotePecPort) {
-    ports.put(pecId, remotePecPort);
+  public void addRemotePecPort(String pecId, PecInnerPortProxy pecInnerPortProxy) {
+    pecPortProxyMap.put(pecId, pecInnerPortProxy);
   }
 
   public PecInnerPort getOrCreateInnerPort(String pecId, String sessionId) {
-    PecMessageReceiver port = ports.get(pecId);
+    PecInnerPort port = pecPortMap.get(pecId);
     if (port == null) {
       return createInnerPecPort(pecId, sessionId);
-    } else if (port instanceof PecInnerPort) {
-      return (PecInnerPort) port;
     } else {
-      throw new IllegalArgumentException("PEC with ID %s is not an inner port: " + pecId);
+      return port;
     }
   }
 
   public void removePecPort(String pecId) {
-    ports.remove(pecId);
+    pecPortMap.remove(pecId);
```

Comment:
let's split these into 2 methods - `remotePecInnerPort` and `remotePecInnerPortProxy`.
also i noticed that we don't call the method from ArcsAndroid, when stopping the arc. 
sorry, this is unrelated to this refactoring, but is a bug, and i'm glad we're fixing it.

Reply:
i also think each of the methods should throw an exception, if the port being removed doesn't exist. thanks!

Reply:
Done.